src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -181,234 +209,248 @@ private void CreateStreamAndStartWatcher()
+                    // Notify the StartRaisingEvents call that we are initialized and about to start
+                    // so that it can return and avoid a race-condition around multiple threads calling Stop and Start
+                    runLoopStarted.Set();
 
-            // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-            // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-            List<FSEventStreamEventId> handledRenameEvents = null;
+                    if (started)
+                    {
+                        // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                        Interop.RunLoop.CFRunLoopRun();
 
-            for (long i = 0; i < numEvents.ToInt32(); i++)
-            {
-                // We need to special case the root path for pattern matching, so cache the path this way in case it is the root
-                // and use it for pattern matching. We'll go back to the eventPaths[i] to get the actual path for notifications
-                string path = eventPaths[i];
-                if (path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
-                {
-                    path += "/.";
+                        // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                        Interop.EventStream.FSEventStreamStop(_eventStream);
+                    }
+                    else
+                    {
+                        // Try to get the Watcher to raise the error event; if we can't do that, just silently exist since the watcher is gone anyway
+                        FileSystemWatcher watcher;
+                        if (_weakWatcher.TryGetTarget(out watcher))
+                        {
+                            // An error occurred while trying to start the run loop so fail out
+                            watcher.OnError(new ErrorEventArgs(new IOException(SR.EventStream_FailedToStart, Marshal.GetLastWin32Error())));
+                        }
+                    }
                 }
-
-                // First, we should check if this event should kick off a re-scan since we can't really rely on anything after this point if that is true
-                if (ShouldRescanOccur(eventFlags[i]))
+                finally
                 {
-                    OnError(new ErrorEventArgs(new IOException(SR.FSW_BufferOverflow, (int)eventFlags[i])));
-                    break;
+                    // Always unschedule the RunLoop before cleaning up
+                    Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
```

Comment:
I guess my previous suggestion of setting _watcherRunLoop to IntPtr.Zero doesn't work because the value is needed here?